libmozdata/vcs_map.py

```diff
@@ -0,0 +1,112 @@
+import argparse
+import os
+import sys
+import urllib.request
+
+import hglib
+import requests
+from tqdm import tqdm
+
+MAPPER_SERVICE = "https://mapper.mozilla-releng.net"
+VCS_MAP_FULL_PATH = "vcs_map_full"
+VCS_MAP_CACHE_PATH = "vcs_map_cache"
+
+git_to_mercurial_mapping = {}
+mercurial_to_git_mapping = {}
+
+
+def download_mapfile():
+    if not os.path.exists(VCS_MAP_FULL_PATH):
+        urllib.request.urlretrieve(
+            f"{MAPPER_SERVICE}/gecko-dev/mapfile/full", VCS_MAP_FULL_PATH
+        )
+
+
+MAPFILE_LOADED = False
+
+
+def load_mapfile():
+    global MAPFILE_LOADED
+    if MAPFILE_LOADED:
+        return
+
+    MAPFILE_LOADED = True
+
+    def _read_mapping_file(path):
+        with open(path, "r") as f:
+            for line in f:
+                git_hash, mercurial_hash = line.rstrip("\n").split(" ")
+                git_to_mercurial_mapping[git_hash] = mercurial_hash
+                mercurial_to_git_mapping[mercurial_hash] = git_hash
+
+    try:
+        _read_mapping_file(VCS_MAP_FULL_PATH)
+    except FileNotFoundError:
+        pass
+
+    try:
+        _read_mapping_file(VCS_MAP_CACHE_PATH)
+    except FileNotFoundError:
+        pass
+
+
+def _write_result(git_hash, mercurial_hash):
+    with open(VCS_MAP_CACHE_PATH, "a") as f:
+        mercurial_to_git_mapping[mercurial_hash] = git_hash
+        git_to_mercurial_mapping[git_hash] = mercurial_hash
+
+        f.write(f"{git_hash} {mercurial_hash}\n")
+
+
+def mercurial_to_git(mercurial_hash):
+    load_mapfile()
+
+    if mercurial_hash not in mercurial_to_git_mapping:
+        r = requests.get(f"{MAPPER_SERVICE}/gecko-dev/rev/hg/{mercurial_hash}")
+        r.raise_for_status()
+
+        git_hash = r.text.split(" ")[0]
+
+        _write_result(git_hash, mercurial_hash)
+
+        return git_hash
+
+    return mercurial_to_git_mapping[mercurial_hash]
+
+
+def git_to_mercurial(git_hash, cache_only=False):
+    load_mapfile()
+
+    if git_hash not in git_to_mercurial_mapping:
+        if cache_only:
+            print(f"missing git commit in the vcs_map: {git_hash}", file=sys.stderr)
```

Comment:
I wonder if it'd better or not to throw an exception (the same for the few return "" below)

Reply:
It's the same for me, but I've done it in 30395e49b415db191da054dc20e68e1bab46a686.